File: src/sequence.cpp
#include "sequence.hpp"
#include "languages/Bair/BairWitness.hpp"
#include <cassert>

namespace libstark {

size_t sequence_usize_get_index(const SequenceUsize &self, size_t idx) {
  return self.seq.getElementByIndex(idx);
}

size_t sequence_usize_size(const SequenceUsize &self) {
  return self.seq.size();
}

// // cxx is not able to return std::vector by value, so we allocate
// // Vec in Rust, pass its reference to C++, then C++ fills it
// // with data it wants to return.
// std::shared_ptr<std::vector<Algebra::FieldElement>> sequence_color_get_index(const SequenceColor &self, size_t idx) {
//   auto val = self.seq.getElementByIndex(idx);
//   return std::make_shared<decltype(val)>(std::move(val));
// }

size_t sequence_color_size(const SequenceColor &self) {
  return self.seq.size();
}

} // namespace libstark

File: src/sequence.rs
use crate::glue::libstark::{sequence_usize_get_index, sequence_usize_size, SequenceUsize};

pub trait Sequence<T> {
    fn get_element_at_index(&self, idx: usize) -> T;
    fn size(&self) -> usize;
}

impl Sequence<usize> for SequenceUsize {
    fn get_element_at_index(&self, idx: usize) -> usize {
        return sequence_usize_get_index(&self, idx);
    }

    fn size(&self) -> usize {
        return sequence_usize_size(&self);
    }
}

impl<'a> IntoIterator for &'a SequenceUsize {
    type Item = usize;
    type IntoIter = SequenceUsizeIterator<'a>;

    fn into_iter(self) -> Self::IntoIter {
        SequenceUsizeIterator {
            seq: &self,
            index: 0,
        }
    }
}

pub struct SequenceUsizeIterator<'a> {
    seq: &'a SequenceUsize,
    index: usize,
}

impl Iterator for SequenceUsizeIterator<'_> {
    type Item = usize;
    fn next(&mut self) -> Option<usize> {
        if self.index >= 0 && self.index < self.seq.size() {
            let value = self.seq.get_element_at_index(self.index);
            self.index += 1;
            Some(value)
        } else {
            None
        }
    }
}
